sql/hive/src/main/java/org/apache/hadoop/hive/ql/io/orc/SparkVectorizedOrcRecordReader.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
+import org.apache.hadoop.hive.serde2.typeinfo.DecimalTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.PrimitiveTypeInfo;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.mapred.RecordReader;
+
+/**
+ * A mapred.RecordReader that returns VectorizedRowBatch.
+ */
+public class SparkVectorizedOrcRecordReader
+      implements RecordReader<NullWritable, VectorizedRowBatch> {
+    private final org.apache.hadoop.hive.ql.io.orc.RecordReader reader;
+    private final long offset;
+    private final long length;
+    private float progress = 0.0f;
+    private ObjectInspector objectInspector;
+
+    SparkVectorizedOrcRecordReader(Reader file, Configuration conf,
+        FileSplit fileSplit) throws IOException {
+      this.offset = fileSplit.getStart();
+      this.length = fileSplit.getLength();
+      this.objectInspector = file.getObjectInspector();
+      this.reader = OrcInputFormat.createReaderFromFile(file, conf, this.offset,
+        this.length);
+      this.progress = reader.getProgress();
+    }
+
+    /**
+     * Create a ColumnVector based on given ObjectInspector's type info.
+     *
+     * @param inspector ObjectInspector
+     */
+    private ColumnVector createColumnVector(ObjectInspector inspector) {
+      switch(inspector.getCategory()) {
+        case PRIMITIVE:
+          {
+            PrimitiveTypeInfo primitiveTypeInfo =
+              (PrimitiveTypeInfo) ((PrimitiveObjectInspector)inspector).getTypeInfo();
+            switch(primitiveTypeInfo.getPrimitiveCategory()) {
+              case BOOLEAN:
+              case BYTE:
+              case SHORT:
+              case INT:
+              case LONG:
+              case DATE:
+              case INTERVAL_YEAR_MONTH:
+                return new LongColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+              case FLOAT:
+              case DOUBLE:
+                return new DoubleColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+              case BINARY:
+              case STRING:
+              case CHAR:
+              case VARCHAR:
+                BytesColumnVector column = new BytesColumnVector(VectorizedRowBatch.DEFAULT_SIZE);
+                column.initBuffer();
+                return column;
+              case DECIMAL:
+                DecimalTypeInfo tInfo = (DecimalTypeInfo) primitiveTypeInfo;
+                return new DecimalColumnVector(VectorizedRowBatch.DEFAULT_SIZE,
+                    tInfo.precision(), tInfo.scale());
+              default:
+                throw new RuntimeException("Vectorizaton is not supported for datatype:"
+                    + primitiveTypeInfo.getPrimitiveCategory());
+            }
+          }
+        default:
+          throw new RuntimeException("Vectorization is not supported for datatype:"
```

Comment:
OK. Let's do the later one. Actually OrcFileFormat has the check for the data types not supported as you commented below. When we come to here, it might be hard to fallback to non-vectorised codepath.